In [29]:
import numpy as np
import pandas as pd

<h1> Boostrap Aggregation </h1>

Boostrap Aggregation also known as bagging, is the ensemle learning method that is commonly used to reduce variacle withing noisy dataset.

<h2> Boostrap Sampling </h2>


In statistics, Bootstrap Sampling is a method that involves drawing of sample data repeatedly with replacement from a data source to estimate a population parameter.

* Sampling: With respect to statistics, sampling is the process of selecting a subset of items from a vast collection of items (population) to estimate a certain characteristic of the entire population
* Sampling with replacement: It means a data point in a drawn sample can reappear in future drawn samples as well
* Parameter estimation: It is a method of estimating parameters for the population using samples. A parameter is a measurable characteristic associated with a population. For example, the average height of residents in a city, the count of red blood cells, etc.

Bootstrap sampling is used in a machine learning ensemble algorithm called bootstrap aggregating (also called bagging). It helps in avoiding overfitting and improves the stability of machine learning algorithms.

let's have an array with standart deviation 3 and mean 5

In [174]:
x = np.random.normal(loc=4, scale=10, size=100)
print(x.mean())

4.035089509803418


let's add outliers to our data set 100 and 50 ,50 

In [175]:
x = np.append(x,[30,50,60])

x

array([ -7.22514737,   9.64008242,  12.46973029,  14.16398874,
        -3.02322046,   6.08719057,  10.38087864,  -1.19306273,
        21.85138381,  23.07221331, -14.83884207,  -7.78851753,
       -23.69885823,  -2.2125994 ,   3.33910124,  -2.71109915,
        14.95556448,   7.12456188,  -8.44276263,  12.01226031,
        27.65845131,   5.40105553,   3.39251665,  31.80208122,
        -0.69864089,  -6.89269376,   0.20402272,  12.047549  ,
         2.13207771,  14.13790503,  10.05622165,   0.79688928,
        -5.60171298,   0.3554777 , -12.42728644,   3.86551729,
       -12.00544461,   9.10785967,  23.58839493,  -4.709603  ,
        -6.26837947,  17.07055526, -27.20611533,   6.64699572,
        20.70892682,  13.0813861 ,   4.51187944,  -5.04844924,
         3.75060747,  12.11225492,  14.22379131,  -5.22917976,
        11.89972929,   2.8388446 ,  17.31709796,  10.41282897,
        -6.88252979,   3.03103921,  -6.48564297,  -7.30055087,
        -1.49784677,  28.47009415,  -8.95626414,  17.23

In [176]:
print(x.mean())
print(x.std())


5.276785931847978
13.567399339893194


we can see that the mean values has been changed from 4,95 to 10.41,
the number 10.41 is wrnong it comes from the noise data [140,50,60].Is there method to achieve mean value wich is closer to 5,in the other word to avoid noisy data [140,50,60] as well as posible.

In [190]:
means =  []
for _ in  range(5610):
    respampling = np.random.choice(x,size=40,replace=True)
    means.append(respampling.mean())
    print(respampling.mean())

6.391154677065265
1.6534205851819739
5.8630292268153035
5.863011029914129
2.941373775853519
5.147701298520902
5.531321424702087
5.166047253919745
7.180920519966591
7.213178035801225
5.1516421632683596
5.095153349320141
5.854319762658295
3.994397235943617
7.172985535407196
3.5080369201648622
0.8178874431117533
8.01419973449705
7.46278132515076
7.108913711158121
4.035227710901415
6.679135458973832
5.2191706300266665
4.406010937577664
6.40601767990029
3.0766249944156763
5.940810424096492
7.485580352208274
5.639395803177512
3.5827054331753097
7.63868406840818
4.656178434301049
8.449362439701892
4.4844277684041405
5.104505836405526
6.805237534733888
1.5735499647886708
3.3870267079172294
4.555973806802872
5.477341411012179
8.69718435408698
4.70555644073391
10.67870017997671
4.199802267150023
5.639674179032855
10.642226003291887
6.33425682891628
7.958651399883882
4.497842808391976
8.111941260085699
6.745481774197687
2.8713369209390125
5.620153679595658
4.681287053480419
5.468217449269469
6.94

6.357561390455418
4.463755473803243
6.016192043099486
9.45326659241202
7.7076167178491675
5.50476850418647
5.662439229875948
2.4007392327990353
5.5308697087155005
3.063286676806462
4.421702395112488
5.996999376962573
1.8657087571870263
7.878654330453495
3.447875907244737
3.683426099383508
5.122032364552302
4.411813321495535
8.706602337929741
2.88466565369243
8.41795196602355
5.89416820102477
8.0687062337447
6.046237540177674
9.150472711187009
4.26397188748654
4.244012098714092
5.421758156869198
7.0761985480689304
3.5247291150341162
3.407960507399028
7.52883038237403
4.972571409947773
4.395819542419568
4.77297271566175
5.242219761056682
3.4075525729607037
4.7897838665913826
8.856507343288499
5.435718281604506
8.603352056936663
1.3700942079554945
10.402251965316617
7.049096625112734
3.6684336798989206
8.816864801048448
5.022852420665631
3.9230708884938084
6.380272232302182
5.36832402297112
8.527827096411034
6.794516329041942
5.740930030601971
4.992064422328413
6.403833246860697
4.1937553

3.4960384108983966
4.625805982417001
2.5193779559133356
6.243589559377213
6.313197986231863
2.9093587646484544
5.660114343690101
3.8810789647998734
4.7270912871148925
4.595908599416047
2.2020457188967675
4.824854254201988
4.32849779718981
5.6067321985191585
-1.467797343374526
2.4614609816790396
4.03857753565863
4.171859211181839
4.6129751311618925
9.370344624271073
2.4933443098992885
7.366407579684356
1.8384166081641453
2.9347059801345465
8.088528910973924
5.186297348124614
6.617866837151718
4.18448493370371
4.461051503129964
7.235662235544457
3.060664003716382
7.114435550163336
8.40843642601793
4.395977859093575
6.379600599928218
7.597709153531701
5.25322729751628
3.0134100797146575
5.454380529168914
7.609697341643408
6.617341877940845
5.192524437068164
8.861447518016597
4.723919861220555
4.704984440698928
6.013483496019967
1.8604597092322561
4.885905858456683
3.121193233797232
4.170726211685301
8.121715566399814
3.5680834892385747
5.204156616727973
7.568837128096059
5.138452432089329

In [191]:
np.array(means).mean()

5.2726753792566265

<h2> References </h2>

[1] <a href='https://www.analyticsvidhya.com/blog/2020/02/what-is-bootstrap-sampling-in-statistics-and-machine-learning/'>What is Bootstrap Sampling in Statistics and Machine Learning </a>